pkg/limits/playback_manager.go

```diff
@@ -0,0 +1,144 @@
+package limits
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/coder/quartz"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/backoff"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/twmb/franz-go/pkg/kgo"
+
+	"github.com/grafana/loki/v3/pkg/limits/proto"
+)
+
+// A PlaybackManager processes records from the metadata topic. It is
+// responsible for replaying newly assigned partitions and merging records
+// from other zones.
+type PlaybackManager struct {
+	client           *kgo.Client
+	partitionManager *PartitionManager
+	usage            *UsageStore
+	// readinessCheck checks if a waiting or replaying partition can be
+	// switched to ready.
+	readinessCheck   PartitionReadinessCheck
+	logger           log.Logger
+	recordsProcessed *prometheus.CounterVec
+
+	// Used for tests.
+	clock quartz.Clock
+}
+
+// NewPlaybackManager returns a new PlaybackManager.
+func NewPlaybackManager(
+	client *kgo.Client,
+	partitionManager *PartitionManager,
+	usage *UsageStore,
+	readinessCheck PartitionReadinessCheck,
+	logger log.Logger,
+	reg prometheus.Registerer,
+) *PlaybackManager {
+	return &PlaybackManager{
+		client:           client,
+		partitionManager: partitionManager,
+		usage:            usage,
+		readinessCheck:   readinessCheck,
+		logger:           logger,
+		clock:            quartz.NewReal(),
+		recordsProcessed: promauto.With(reg).NewCounterVec(
+			prometheus.CounterOpts{
+				Name: "loki_ingest_limits_records_processed_total",
+				Help: "The total number of records processed.",
+			},
+			[]string{"partition"},
+		),
+	}
+}
+
+func (m *PlaybackManager) Run(ctx context.Context) {
+	b := backoff.New(ctx, backoff.Config{
+		MinBackoff: 100 * time.Millisecond,
+		MaxBackoff: time.Second,
+		MaxRetries: 0,
+	})
+	for b.Ongoing() {
```

Comment:
What is the use case for backoff on poll fetches here?

Reply:
This is quite subtle, but it's when `PollFetches` returns immediately because of a failure. It can be because the broker is unreachable/down, or a configuration error (as I saw while testing this in dev 😄). But tl;dr is it stops us calling the function in a tight loop and burning CPU.